packages/@n8n/task-runner-python/src/logs.py

```diff
@@ -70,3 +76,23 @@ def setup_logging():
     logger.addHandler(stream_handler)
 
     logging.getLogger("websockets.client").setLevel(logging.DEBUG)
+
+
+def format_print_args(*args) -> list[str]:
+    """Format `print` arguments for browser console output, handling circular references."""
+
+    formatted_args = []
+    for arg in args:
+        if isinstance(arg, dict) and CIRCULAR_REFERENCE_PLACEHOLDER_KEY in arg:
+            formatted_args.append(f"[Circular {arg['__type__']}]")
+        elif isinstance(arg, str):
+            formatted_args.append(f"'{arg}'")
+        elif isinstance(arg, (int, float, bool)) or arg is None:
+            formatted_args.append(str(arg))
+        else:
+            try:
+                formatted_args.append(json.dumps(arg, default=str, ensure_ascii=False))
+            except Exception as _:
+                formatted_args.append(f"[Circular {type(arg).__name__}]")
```

Comment:
Are we certain the error is always due to circular ref, and not something else? Can we identify it?

Reply:
Refactored this to simplify.